rdr_service/tools/tool_libs/backfill_questionnaire_answers.py

```diff
@@ -0,0 +1,169 @@
+from rdr_service.lib_fhir.fhirclient_1_0_6.models import questionnaireresponse as fhir_questionnaireresponse
+from rdr_service.model.code import CodeType
+from rdr_service.model.questionnaire_response import QuestionnaireResponseAnswer, QuestionnaireResponse
+from rdr_service.tools.tool_libs.tool_base import cli_run, ToolBase
+
+tool_cmd = 'questionnaire-answers-backfill'
+tool_desc = ('Backfill the questionnaire answers table for questionnaires that are missing a concept code definition '
+             '- Created for DA-4318')
+
+
+class QuestionnaireAnswersBackfillTool(ToolBase):
+    def run(self):
+        super(QuestionnaireAnswersBackfillTool, self).run()
+
+        questionnaire_ids = [self.args.questionnaire_ids]
+        latest_id = -10
+        with self.get_session() as session:
+            found_responses = True
+            while found_responses:
+                questionnaire_response_query = session.query(
+                    QuestionnaireResponse.questionnaireResponseId,
+                    QuestionnaireResponse.resource,
+                    QuestionnaireResponse.questionnaireId,
+                ).filter(
+                    QuestionnaireResponse.questionnaireResponseId > latest_id,
+                    QuestionnaireResponse.created > '2021-11-01',
+                    QuestionnaireResponse.questionnaireId.in_(questionnaire_ids),
+                    QuestionnaireResponse.classificationType == 0,
+                    QuestionnaireResponse.resource.like('%linkId": "60666%')
+                ).order_by(QuestionnaireResponse.questionnaireResponseId).limit(500)
+
+                found_responses = False
+                for response_data in questionnaire_response_query:
+                    found_responses = True
+
+                    resource_json = json.loads(response_data.resource)
+                    questionnaire_id = response_data.questionnaireId
+                    fhir_qr = fhir_questionnaireresponse.QuestionnaireResponse(resource_json)
+                    questionnaire = self._get_questionnaire(questionnaire_id)
+
+                    if fhir_qr.group is not None:
+                        # Extract answers from resource json
+                        code_map, answers = self._extract_codes_and_answers(fhir_qr.group, questionnaire)
+
+                        # Get or insert codes, and retrieve their database IDs.
+                        code_id_map = CodeDao().get_internal_id_code_map(code_map)
+
+                        # Now add the child answers, using the IDs in code_id_map
+                        answers_to_insert = self._add_answers(code_id_map, answers,
+                                                              response_data.questionnaireResponseId)
+                        session.add_all(answers_to_insert)
+
+                    latest_id = response_data.questionnaireResponseId
+
+                if found_responses:
+                    print(f'got to {latest_id}')
```

Comment:
May want to move these print statements down after the `commit`, to confirm it was successful and nothing got rolled back?

If the tool script fails at any point (can happen with long-running scripts due to failed network connections, etc.), I assume before any subsequent run, the starting value of `latest_id` will be changed in the code and instead of -10, will be set to the last value printed out on the previous run.   Need to make sure the same `questionnaire_response_id` can't be processed by this script twice, or we may end up with duplicate/extra `questionnaire_response_answer` records.

Reply:
Good call out! I will update to make sure we don't reprocess any `questionnaire_response_ids`